Enable `use feature 'module_true'` (reworked squash)

demerphq opened the pull request:
Per RFC 18, whenever `use feature 'module_true';` is enabled in a scope, any file required with `require` has an implicit return value of true and will not trigger the "did not return a true value" error condition.

See also https://github.com/Perl/perl5/pull/20365 which this PR is designed to replace. I have taken the liberty of pushing this patch with @chromatic as the author. I just tinkered with the implementation.

chromatic committed:
Enable `use feature 'module_true'`

Per RFC 18, whenever `use feature 'module_true';` is enabled in a scope,
any file required with `require` has an implicit return value of true
and will not trigger the "did not return a true value" error condition.

This includes logic to use the OPf_SPECIAL flag for OP_RETURN listops to
indicate that the module_true feature is in effect when it executes.
This flag plays no role unless the OP_RETURN tail calls the pp_leaveeval
logic, so it doesn't affect normal sub returns.

demerphq commented:
This resolves the issues raised by @tonycoz in his review of an earlier version of the code: https://github.com/Perl/perl5/commit/d6b8be71272aafbbaa17bb27be4ac484f8726651 and all the known issues raised in https://github.com/Perl/perl5/pull/20365

With this version of the PR we test under the debugger and not, and all the permutations and different ways the code might present.

@chromatic, please review again, and I think we need a perldelta. Will you pick that up?

@leonerd any further thoughts?

demerphq commented on t/comp/require.t:
```diff
@@ -348,6 +354,103 @@ if (defined &DynaLoader::boot_DynaLoader) {
+}
+
+{
+    BEGIN { ${^OPEN} = ":utf8\0"; }
```
this is cargo culted all over this test file. i really wonder if it should be.

chromatic replied:
It's neither necessary for nor relevant to the tests I added. It was 100% cargo culted in those tests and should be safe to remove.

demerphq replied:
Removed with the last squash/push.

demerphq commented:
I pushed a couple of WS/comment tweaks that should be squashed down before it is merged.

chromatic commented:
I have a couple of additional test cases I want to run this through (specifically `no feature 'module_true';`; I'll follow up hopefully before tomorrow with my results.

tonycoz commented:
The require entry in `pod/perlfunc.pod` probably needs an update too.

tonycoz commented:
From the RFC:
```
tony@venus:.../git/perl5$ cat >Demo2.pm
package Demo2;
use feature 'module_true';

return 1 if $main::test_1;
return 0 if $main::test_2;

{
  no feature 'module_true';
  return 0 if $main::test_3;
}
tony@venus:.../git/perl5$ ./perl -Ilib -I. -e '$test_3 = 1; require Demo2;'
tony@venus:.../git/perl5$ 
```
According to the RFC this should fail.

I suspect to implement what the RFC requires setting a flag on OP_REQUIRE and looking for that in pp_leaveeval (pp_return would ignore it).

tonycoz commented on pp_ctl.c:
```diff
@@ -4845,8 +4846,54 @@ PP(pp_leaveeval)
+             * or find something that is true. */
+            if (check) {
+                const OP *kid = cLISTOPx(check)->op_first;
+                if (0) op_dump(kid);
```
leftover debugging code, I assume

demerphq replied:
Yep, will remove.

demerphq replied:
Well, maybe not yet, its still useful for debugging. :-)

chromatic commented:
Tony hit one of the test cases I had thought of. Here are the other few I thought were material:

```
diff --git a/t/comp/require.t b/t/comp/require.t
index 05c3270f9e..7fb585df88 100644
--- a/t/comp/require.t
+++ b/t/comp/require.t
@@ -415,6 +415,9 @@ BEGIN {
             }
         }
     }
+
+    # and more later on
+    $module_true_test_count += 4;
 }
 
 {
@@ -448,6 +451,34 @@ BEGIN {
                   . "is simple true value <$return_val[0]>\n";
         }
     }
+
+    {
+        write_file('bleah.pm', "package bleah;\nuse v5.37;\nsub foo {};\nno feature 'module_true';\n");
+
+        local $@;
+        my $result = 0;
+        my $not = eval "\$result = require 'bleah.pm'; 1" ? 'not ' : '';
+        $i++;
+        print "${not}ok $i - disabling module_true should not return a true value ($result)\n";
+        $not = $@ =~ /bleah.pm did not return a true value/ ? '' : 'not ';
+        $i++;
+        print "${not}ok $i - ... and should fail to compile without a true return value\n";
+    }
+
+    {
+        write_file('bleah.pm', "package bleah;\nuse feature ':5.38';\nsub bar {};\nno feature 'module_true';\n3;\n");
+
+        local $@;
+        my $result = 0;
+        eval "\$result = require 'bleah.pm'; 1";
+        $not = $result == 3 ? '' : 'not ';
+        $i++;
+        print "${not}ok $i - disabling 'module_true' and should not override module's return value ($result)\n";
+        $not = $@ eq '' ? '' : 'not ';
+        $i++;
+        print "${not}ok $i - ... but should compile successfully with a provided return value ($@)\n";
+    }
+
 }
 
 ##########################################
```

demerphq commented:
Ill follow up on the test cases. @chromatic be careful with reusing "bleah" like that, we are installing subs in the name space and require populates %INC, and even if we clear %INC so we can re-require the file sub foo{} stays in the namespace and we get redefined warnings from the test.  Anyway, ill massage these tests into place and see how it goes.

demerphq commented:
@tonycoz and @chromatic the failing cases involve optrees like this:

```
736  leaveeval UNOP(0x559b1328bb80) ===> [0x0]
     PARENT ===> [0x0]
     TARG = 1
     FLAGS = (SCALAR,KIDS,SLABBED,SAVEFREE)
     PRIVATE = (REFC,0x1)
     REFCNT = 1
     |   
737  +--lineseq LISTOP(0x559b1328bbb8) ===> 736 [leaveeval 0x559b1328bb80]
         FLAGS = (SCALAR,KIDS,SLABBED,SAVEFREE)
         |   
296      +--stub OP(0x559b1328bc58) ===> 736 [leaveeval 0x559b1328bb80]
         |   FLAGS = (SCALAR,SLABBED,MORESIB)
         |   
738      +--null (ex-nextstate) COP(0x559b1328bbf8) ===> 737 [lineseq 0x559b1328bbb8]
             FLAGS = (SCALAR,SLABBED)
             LINE = 3
             PACKAGE = "blunge"
             SEQ = 117
```

You can see what has happened, the optimizer has kicked in and converted the nextstate op we need to find the feature flags into a NULL op. I will see if i can find what does that and make it not do it if there are feature flags set.

demerphq commented:
@tonycoz @chromatic actually, scratch that last comment. It makes no sense. We are mistakenly thinking the feature is still enabled, when it isn't, so it cant be a problem finding the nextstate node.

demerphq committed:
dump.c - in op_dump show the hints and feature bits

I used to 'FEATS' instead of 'FEATURES' so the values line up
nicely.

demerphq committed:
op.c - remove superfluous whitespace

demerphq committed:
peep.c - add a comment/question to a comment

We optimize away certain nextstate ops, but it isn't clear if
it is possible for the ops to be there to switch hints or feature
flags. Tony thinks its fine however. So I have added the commentary
to the comment for future readers.

demerphq committed:
perlfunc.pod - document module_true in the require entry

A brief note pointing people at the feature manpage.

demerphq commented:
@tonycoz @chromatic: I have updated the code to use OPf_SPECIAL on OP_RETURN listops to avoid the need to grub around in the optree for return, as @tonycoz predicted it is otherwise really hard to find the the right COP to look at.

However there was a logic error as well in the previous code for finding the nextstate operator, which is why the "no feature" cases were failing sometimes, we were using the first NEXTSTATE op and we should have been using the last. Fixed now.

Added a note to perlfunc for require as requested, and merged the tests you guys raised as well.

chromatic commented:
I've tested this and it looks good to me. No further comments.

demerphq commented:
@tonycoz any further considerations?

tonycoz commented on pod/perlfunc.pod:
```diff
@@ -6746,11 +6746,14 @@ would have semantics similar to the following:
 end such a file with C<1;> unless you're sure it'll return true
 otherwise.  But it's better just to put the C<1;>, in case you add more
-statements.
+statements. As of 5.37.6 this requirement may be avoided by enabling
+the 'module_true' feature, which is enabled by default in modern
+version bundles. Thus code with C<use v5.37;> no longer needs to concern
+itself with this issue. See L<feature> for more details.
```
perhaps this should mention you want to enable module_true in the required file, not in the scope that calls require.  This is mentioned in feature.pm, but I think it should be mentioned here.

demerphq replied:
Done.

tonycoz commented on peep.c:
```diff
@@ -3032,7 +3032,9 @@ Perl_rpeep(pTHX_ OP *o)
                to carry two labels. For now, take the easier option, and skip
-               this optimisation if the first NEXTSTATE has a label.  */
+               this optimisation if the first NEXTSTATE has a label.
+               XXX: Yves wonders what about if they have different hints or features?
+               */
```
The code removes the first COP, not the second, so I think that's safe.

demerphq replied:
Comment changed to note your reply.

demerphq commented:
@tonycoz you good with this? Unless I hear an objection ill merge this tomorrow.

demerphq commented:
Thanks @chromatic for starting this PR. I learned a lot helping getting it wrapped up. Sorry for hijacking your PR, but the main patch is credited to you, and you will be in the perl authors list for this release as well.

leonerd commented:
Woo \o/ Excellent work, all. Much thanks.

chromatic commented:
No apologies necessary @demerphq . It's a team effort, much better results than if I'd did it myself. It was a pleasure to collaborate.
